include/xgboost/tree_model.h

```diff
@@ -184,127 +193,104 @@ class TreeModel {
     unsigned sindex_{0};
     // extra info
     Info info_;
-    // set parent
-    inline void SetParent(int pidx, bool is_left_child = true) {
-      if (is_left_child) pidx |= (1U << 31);
-      this->parent_ = pidx;
-    }
   };
 
- protected:
+ private:
   // vector of nodes
   std::vector<Node> nodes_;
-  // free node space, used during training process
-  std::vector<int>  deleted_nodes_;
   // stats of nodes
-  std::vector<TNodeStat> stats_;
-  // leaf vector, that is used to store additional information
-  std::vector<bst_float> leaf_vector_;
+  std::vector<RTreeNodeStat> stats_;
+
+  std::vector<bst_float>
+      node_mean_values_;  // Cache mean values for generating contributions
   // allocate a new node,
   // !!!!!! NOTE: may cause BUG here, nodes.resize
   inline int AllocNode() {
-    if (param.num_deleted != 0) {
```

Comment:
This is the logic for recycling pruned nodes and need to be kept here

Reply:
I put an exit(-1) statement inside this and ran all of the Python tests without any issues. Unless I'm missing something, these recycled nodes never get used again. The pruner deletes nodes but there is nothing that adds nodes after pruning has occurred. 

The reason for removing this is to reduce the number of member variables and statefullness of the tree structure. This will make it more flexible in future.

Reply:
This will only be triggered if certain nodes are pruned and alloc reuse them in the future tree builds. I think it is still an useful feature to have here.